"`pencilSketch` not declared in scope" error on install

When installing a compilation error appears, which I can replicate when trying to compile directly:

```bash
$  /opt/rh/devtoolset-7/root/usr/bin/g++ -m64 -std=gnu++11 -I"/usr/include/R" -DNDEBUG -I/usr/include/opencv   -D_DATA_PREFIX=/usr -I"/usr/lib64/R/library/Rcpp/include" -I/usr/local/include  -fpic  -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic  -c effects.cpp -o effects.o
effects.cpp: In function ‘XPtrMat cvmat_sketch(XPtrMat, bool)’:
effects.cpp:16:3: error: ‘pencilSketch’ was not declared in this scope
   pencilSketch(get_mat(ptr),out1, out2, 10 , 0.1f, 0.03f);
   ^~~~~~~~~~~~
effects.cpp: In function ‘XPtrMat cvmat_stylize(XPtrMat)’:
effects.cpp:23:3: error: ‘stylization’ was not declared in this scope
   stylization(get_mat(ptr), out);
   ^~~~~~~~~~~
```

Is there a specific version of `opencv` I should be using?

`opencv` version:

```bash
$  sudo yum list installed \| grep opencv
opencv.x86_64                      2.4.5-3.el7                         @base
opencv-core.x86_64                 2.4.5-3.el7                         @base
opencv-devel.x86_64                2.4.5-3.el7                         @base
opencv-python.x86_64               2.4.5-3.el7                         @base
```
`R` version:

```bash
$  sudo yum list installed | grep ^R
R.x86_64                           3.6.0-1.el7                         @epel
R-Rcpp.x86_64                      1.0.2-1.el7                         @epel
R-Rcpp-devel.x86_64                1.0.2-1.el7                         @epel
R-core.x86_64                      3.6.0-1.el7                         @epel
R-core-devel.x86_64                3.6.0-1.el7                         @epel
R-devel.x86_64                     3.6.0-1.el7                         @epel
R-java.x86_64                      3.6.0-1.el7                         @epel
R-java-devel.x86_64                3.6.0-1.el7                         @epel
```
`g++` version:

```bash
$  /opt/rh/devtoolset-7/root/usr/bin/g++ --version
g++ (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,13 +6,6 @@ pandoc: false
 
 matrix:
   include:
-    - dist: xenial # legacy opencv2
-      addons:
-        apt:
-          update: true
-          packages:
-            - libopencv-dev
-            - opencv-data
     - dist: xenial # up-to-date 3.4 branch
       addons:
         apt:

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,7 +6,16 @@ pandoc: false
 
 matrix:
   include:
+    - dist: xenial # legacy opencv2
     - dist: xenial # up-to-date 3.4 branch
+      addons:
+        apt:
+          update: true
+          sources:
+            - sourceline: 'ppa:cran/opencv'
+          packages:
+            - libopencv-dev
+            - opencv-data
     - dist: xenial # 3.2 branch (Bionic / Buster)
       addons:
         apt:
@@ -29,12 +38,3 @@ matrix:
       osx_image: xcode7.3
       disable_homebrew: true
       before_install: sed -i.bak 's/-isysroot /-I/g' $(R RHOME)/etc/Makeconf
-
-addons:
-  apt:
-    update: true
-    sources:
-      - sourceline: 'ppa:cran/opencv'
-    packages:
-      - libopencv-dev
-      - opencv-data

--- a/.travis.yml
+++ b/.travis.yml
@@ -21,6 +21,9 @@ matrix:
       brew_packages: opencv@3
       env: PKG_CONFIG_PATH="/usr/local/opt/opencv@3/lib/pkgconfig"
     - os: osx
+      brew_packages: opencv@2
+      env: PKG_CONFIG_PATH="/usr/local/opt/opencv@2/lib/pkgconfig"
+    - os: osx
       brew_packages: opencv
     - os: osx
       osx_image: xcode7.3

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,6 +7,12 @@ pandoc: false
 matrix:
   include:
     - dist: xenial # legacy opencv2
+      addons:
+        apt:
+          update: true
+          packages:
+            - libopencv-dev
+            - opencv-data
     - dist: xenial # up-to-date 3.4 branch
       addons:
         apt:

--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -9,7 +9,7 @@ Description: Experimenting with computer vision and machine learning in R. This
     such as edge, body or face detection. These can either be applied to analyze 
     static images, or to filter live video footage from a camera device.
 License: MIT + file LICENSE
-SystemRequirements: OpenCV: libopencv-dev (Debian, Ubuntu) or opencv-devel (Fedora)
+SystemRequirements: OpenCV 3 or newer: libopencv-dev (Debian, Ubuntu) or opencv-devel (Fedora)
 URL: https://docs.ropensci.org/opencv,  https://opencv.org (upstream) https://github.com/ropensci/opencv (devel)
 BugReports: https://github.com/ropensci/opencv/issues
 LinkingTo: Rcpp

--- a/R/init.R
+++ b/R/init.R
@@ -30,7 +30,7 @@ is_check <- function(){
 }
 
 is_mojave <- function(){
-  ver <- utils::tail(strsplit(sessionInfo()$running, ' ')[[1]], 1)
+  ver <- utils::tail(strsplit(utils::sessionInfo()$running, ' ')[[1]], 1)
   as.numeric_version(ver) >= "10.14"
 }
 

--- a/configure
+++ b/configure
@@ -4,7 +4,7 @@ PKG_CONFIG_NAME="opencv4"
 PKG_CONFIG_NAME_ALT="opencv"
 PKG_DEB_NAME="libopencv-dev"
 PKG_RPM_NAME="opencv-devel"
-PKG_BREW_NAME="opencv"
+PKG_BREW_NAME="opencv@2"
 PKG_TEST_HEADER="<opencv2/opencv.hpp>"
 PKG_LIBS="-lopencv_{stitching,superres,videostab,aruco,bgsegm,bioinspired,ccalib,dnn_objdetect,\
 dpm,face,photo,fuzzy,hfs,img_hash,line_descriptor,optflow,reg,rgbd,saliency,stereo,structured_light,\

--- a/src/effects.cpp
+++ b/src/effects.cpp
@@ -13,13 +13,21 @@ XPtrMat cvmat_blur(XPtrMat ptr, size_t ksize){
 XPtrMat cvmat_sketch(XPtrMat ptr, bool color){
   Mat out1;
   Mat out2;
+#if CV_VERSION_EPOCH < 3
+  throw std::runtime_error("pencilSketch requires OpenCV 3 or newer");
+#else
   pencilSketch(get_mat(ptr),out1, out2, 10 , 0.1f, 0.03f);
+#endif
   return cvmat_xptr(color ? out2 : out1);
 }
 
 // [[Rcpp::export]]
 XPtrMat cvmat_stylize(XPtrMat ptr){
   Mat out;
+#if CV_VERSION_EPOCH < 3
+  throw std::runtime_error("stylization requires OpenCV 3 or newer");
+#else
   stylization(get_mat(ptr), out);
+#endif
   return cvmat_xptr(out);
 }

--- a/src/face.cpp
+++ b/src/face.cpp
@@ -60,6 +60,9 @@ XPtrMat cvmat_facemask(XPtrMat ptr, const char * facedata){
 
 // [[Rcpp::export]]
 XPtrMat cvmat_mog2(XPtrMat ptr) {
+#if CV_VERSION_EPOCH < 3
+  throw std::runtime_error("createBackgroundSubtractorMOG2 requires OpenCV 3 or newer");
+#else
   static Ptr<BackgroundSubtractorMOG2> model = createBackgroundSubtractorMOG2();
   model->setVarThreshold(10);
   cv::Mat frame = get_mat(ptr);
@@ -67,15 +70,20 @@ XPtrMat cvmat_mog2(XPtrMat ptr) {
   model->apply(frame, mask);
   //refineSegments(frame, mask, out_frame);
   return cvmat_xptr(mask);
+#endif
 }
 
 // [[Rcpp::export]]
 XPtrMat cvmat_knn(XPtrMat ptr) {
+#if CV_VERSION_EPOCH < 3
+  throw std::runtime_error("createBackgroundSubtractorKNN requires OpenCV 3 or newer");
+#else
   static Ptr<BackgroundSubtractorKNN> model = createBackgroundSubtractorKNN();
   cv::Mat frame = get_mat(ptr);
   cv::Mat mask, out_frame;
   model->apply(frame, mask);
   return cvmat_xptr(mask);
+#endif
 }
 
 // [[Rcpp::export]]

--- a/src/hog.cpp
+++ b/src/hog.cpp
@@ -35,7 +35,8 @@ XPtrMat cvmat_hog(XPtrMat ptr){
     r.width = cvRound(r.width*0.8);
     r.y += cvRound(r.height*0.07);
     r.height = cvRound(r.height*0.8);
-    rectangle(get_mat(ptr), r.tl(), r.br(), cv::Scalar(0,255,0), 3);
+    Mat img = get_mat(ptr);
+    rectangle(img, r.tl(), r.br(), cv::Scalar(0,255,0), 3);
   }
   return ptr;
 }

--- a/src/util.cpp
+++ b/src/util.cpp
@@ -89,34 +89,3 @@ void detectAndDraw( Mat img, CascadeClassifier& cascade,
   }
   //imshow( "result", img );
 }
-
-void refineSegments(const Mat& img, Mat& mask, Mat& dst)
-{
-  int niters = 3;
-  vector<vector<Point> > contours;
-  vector<Vec4i> hierarchy;
-  Mat temp;
-  dilate(mask, temp, Mat(), Point(-1,-1), niters);
-  erode(temp, temp, Mat(), Point(-1,-1), niters*2);
-  dilate(temp, temp, Mat(), Point(-1,-1), niters);
-  findContours( temp, contours, hierarchy, RETR_CCOMP, CHAIN_APPROX_SIMPLE );
-  dst = Mat::zeros(img.size(), CV_8UC3);
-  if( contours.size() == 0 )
-    return;
-  // iterate through all the top-level contours,
-  // draw each connected component with its own random color
-  int idx = 0, largestComp = 0;
-  double maxArea = 0;
-  for( ; idx >= 0; idx = hierarchy[idx][0] )
-  {
-    const vector<Point>& c = contours[idx];
-    double area = fabs(contourArea(Mat(c)));
-    if( area > maxArea )
-    {
-      maxArea = area;
-      largestComp = idx;
-    }
-  }
-  Scalar color( 0, 0, 255 );
-  drawContours( dst, contours, largestComp, color, FILLED, LINE_8, hierarchy );
-}

--- a/src/util.hpp
+++ b/src/util.hpp
@@ -2,7 +2,6 @@
 void detectAndDraw( cv::Mat img, cv::CascadeClassifier& cascade,
                     cv::CascadeClassifier& nestedCascade,
                     double scale, bool tryflip );
-void refineSegments(const cv::Mat& img, cv::Mat& mask, cv::Mat& dst);
 cv::Mat get_mat(XPtrMat image);
 XPtrMat cvmat_xptr(cv::Mat *frame);
 XPtrMat cvmat_xptr(cv::Mat orig);

